pinot-spi/src/main/java/org/apache/pinot/spi/env/PinotConfiguration.java

```diff
@@ -170,8 +173,15 @@ public static List<Configuration> applyDynamicEnvConfig(List<Configuration> conf
       Map<String, String> environmentVariables) {
     return configurations.stream().peek(configuration -> {
       for (String dynamicEnvConfigVarName : configuration.getStringArray(ENV_DYNAMIC_CONFIG_KEY)) {
-        configuration.setProperty(dynamicEnvConfigVarName,
-            environmentVariables.get(configuration.getString(dynamicEnvConfigVarName)));
+        //if the environment variable doesn't exist or the property is already checked do not add the property
+        //TODO: The configuration in controller, server iterating to check and add env variables twice. Need to solve
+        // this better.
+        Object envVar = environmentVariables.get(configuration.getString(dynamicEnvConfigVarName));
+        if (envVar != null) {
```

Comment:
What about saving in the config as a property something like
`templated=true`
and using that to check if applyDynamicEnvConfig should be applied.

Reply:
I mean something like:
```suggestion
        configuration.setProperty(TEMPLATED_KEY, true);
```
This could be red in the next executions from `configuration` to check if the templating has happened already.
Leaving this like this exposes you to issues like loading twice from random keys in the env. For example:
Run 1:
```
property=ENV_VAR_NAME -> Lookup(ENV_VAR_NAME)=path
```
Run 2:
```
property=path -> Lookup(path)=path_value
```

This is unwanted and risky behavior in my opinion.